File: solution/simulation/21610/main.py
# Authored by : gusdn3477
# Co-authored by : -
# Link : http://boj.kr/709edf708bcb49e3972565b8fce5a993

import sys
from collections import deque
from copy import deepcopy

def input():
    return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
arr = []
move = []
visited = [[0 for x in range(N)] for y in range(N)]
storm = []
nx = [0,-1,-1,-1,0,1,1,1]
ny = [-1,-1,0,1,1,1,0,-1]

for i in range(N):
    arr.append(list(map(int, input().split())))

for i in range(M):
    move.append(list(map(int, input().split())))

storm.append((N-1, 0))
storm.append((N-1, 1))
storm.append((N-2, 0))
storm.append((N-2, 1))

for i in range(M):
    d, s = move[i][0]-1, move[i][1]
    visited = [[0 for x in range(N)] for y in range(N)]
    new_storm = []
    for j in range(len(storm)):
        dx = storm[j][0] + nx[d] * s
        dy = storm[j][1] + ny[d] * s
        if dx < 0:
            while True:
                new_dx = dx + N
                if new_dx >= N:
                    break
                dx += N
        elif dx >= N:
            while True:
                new_dx = dx - N
                if new_dx < 0:
                    break
                dx -= N
        if dy < 0:
            while True:
                new_dy = dy + N
                if new_dy >= N:
                    break
                dy += N
        elif dy >= N:
            while True:
                new_dy = dy - N
                if new_dy < 0:
                    break
                dy -= N
        new_storm.append((dx,dy))
        visited[dx][dy] = 1

    for j in range(len(new_storm)):
        arr[new_storm[j][0]][new_storm[j][1]] += 1

    save = []
    for j in range(len(new_storm)):
        ct = 0
        for z in range(1,8,2):
            dx = new_storm[j][0] + nx[z]
            dy = new_storm[j][1] + ny[z]
            if dx < 0 or dx >= N:
                continue
            if dy < 0 or dy >= N:
                continue
            if arr[dx][dy] > 0:
                ct += 1
        save.append(ct)

    for j in range(len(new_storm)):
        arr[new_storm[j][0]][new_storm[j][1]] += save[j]

    storm = []
    for x in range(N):
        for y in range(N):
            if visited[x][y] == 0 and arr[x][y] >= 2:
                arr[x][y] -= 2
                storm.append((x,y))

total = 0
for i in range(N):
    for j in range(N):
        if arr[i][j] > 0:
            total += arr[i][j]

print(total)

File: solution/simulation/21610/main.js
// Authored by : yundosa2
// Co-authored by : -
// Link: http://boj.kr/57673ced91544798ab3cbda0c8cb5525

let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

function solution(input) {
  let [N, M] = input[0].split(" ").map(Number);
  let board = input.slice(1, N + 1).map((v) => v.split(" ").map(Number));
  let moves = input.slice(N + 1).map((v) => v.split(" ").map(Number));
  let clouds = [
    [N - 2, 0],
    [N - 2, 1],
    [N - 1, 0],
    [N - 1, 1],
  ];
  const directions = {
    1: [0, -1],
    2: [-1, -1],
    3: [-1, 0],
    4: [-1, 1],
    5: [0, 1],
    6: [1, 1],
    7: [1, 0],
    8: [1, -1],
  };
  const diagX = [-1, -1, 1, 1];
  const diagY = [-1, 1, -1, 1];
  let answer = 0;

  for (let move of moves) {
    // 1번
    const nextClouds = [];
    for (let cloud of clouds) {
      const [row, col] = cloud;
      const direction = directions[move[0]];
      const distance = move[1];
      let nRow = (row + direction[0] * distance) % N;
      let nCol = (col + direction[1] * distance) % N;
      nRow = nRow < 0 ? nRow + N : nRow;
      nCol = nCol < 0 ? nCol + N : nCol;
      nextClouds.push([nRow, nCol]);
    }

    // 2번
    let visited = Array.from({ length: N }, () => Array(N).fill(false));
    for (let cloud of nextClouds) {
      const [row, col] = cloud;
      board[row][col]++;
      visited[row][col] = true;
    }

    // 3번
    clouds = [];

    // 4번
    for (let cloud of nextClouds) {
      const [row, col] = cloud;
      let count = 0;
      for (let i = 0; i < 4; i++) {
        const [nRow, nCol] = [row + diagX[i], col + diagY[i]];
        if (nRow < 0 || nRow >= N || nCol < 0 || nCol >= N || board[nRow][nCol] < 1) continue;
        count++;
      }
      board[row][col] += count;
    }

    // 5번
    for (let row = 0; row < N; row++) {
      for (let col = 0; col < N; col++) {
        if (board[row][col] >= 2 && !visited[row][col]) {
          board[row][col] -= 2;
          clouds.push([row, col]);
        }
      }
    }
  }

  // 최종 계산
  for (let row = 0; row < N; row++) {
    for (let col = 0; col < N; col++) {
      answer += board[row][col];
    }
  }
  return answer;
}

console.log(solution(input));
